lib/spack/spack/test/concretize_preferences.py

```diff
@@ -416,3 +416,13 @@ def test_multivalued_variants_are_lower_priority_than_providers(self):
         ):
             s = Spec('somevirtual').concretized()
             assert s.name == 'some-virtual-preferred'
+
+    @pytest.mark.regression('26721,19736')
+    def test_sticky_variant_accounts_for_packages_yaml(self):
+        with spack.config.override(
+                'packages:sticky-variant', {
+                    'variants': '+allow-gcc'
+                }
+        ):
+            s = Spec('sticky-variant %gcc').concretized()
+            assert s.satisfies('%gcc') and s.satisfies('+allow-gcc')
```

Comment:
This test checks that we account for `packages.yaml` to override the default in `package.py`

Reply:
Yeah, I meant the other way around.

In my understanding packages.yaml specifies preferences, whereas the abstract spec input specifies constraints.

Now suppose the expressed preference in packages.yaml for the value of a sticky variant would be unsatisfiable, then I would find it consistent if the value toggled back.

So:

```python
class MyPkg:
    variant('x', default=False, sticky=True)
    conflicts('+x')
```
```yaml
packages:
  mypkg:
    variants: ['+x']
```
should concretize to `~x` instead of failing?

The reason being: (a) the stickiness applies to the particular variant default value the package author sets, and (b) if we later add support for discriminating between hard constraints in packages.yaml vs soft preferences (as we currently only have), it'd be nice if everything in packages.yaml was a preference

Reply:
Okay, then I'll stop saying packages.yaml specifies preferences :upside_down_face:

Reply:
I share some of @haampie's concerns here. In all other cases, packages.yaml is a soft preference, I'm not sure it's intuitive that it becomes a hard preference for sticky variants

Reply:
So here's how I see it:
1. The preference is due to the fact that `packages.yaml` overrides the default in `package.py` (even when the variant is not sticky)
2. If the variant is sticky the default value cannot be changed by clingo

The claim in point 1. follows by:

https://github.com/spack/spack/blob/bae2a2660e5c559f0e45eda97502bd2b1dcf3fae/lib/spack/spack/solver/concretize.lp#L485-L501

Furthermore, if we use the semantics currently proposed here the variant can be sticky whether the packager selects the "permissive" or the "restrictive" behavior as a default in `package.py`. Think of having:
```variant
variant('allow-all-compilers', default=True, sticky=True)
``` 
I might want to say:
```yaml
packages:
  foo:
    variants: ~allow-all-compilers
```
to prefer the restrictive default. With the semantic you propose instead that would be impossible from `packages.yaml` (if the permissive semantics is the default in `package.py` clingo would be happy to switch it in case I try to reset it in `packages.yaml`).

Reply:
I can see what @haampie is saying... kind of.  I'm having a hard time coming up with a use case for the soft preference for a sticky variant, though.  It seems inconsistent to me that if the variant is sticky, and you override it, that it would not still be sticky.  I've always thought of `packages.yaml` as overriding preferences and not semantics. 

Do you have a use case for preference semantics for sticky variants @haampie?

With the hard constraints we are proposing in #27987, you'd be able to do:

```yaml
packages:
    all:
        require:[+allow-all-compilers]
```

So we could wait for that.  I do think people will be very confused when they set the value for a sticky variant and it's no longer sticky when they do that.  Also, I can't tell if @haampie expects the soft preferences to only fall back to the sticky default, or if he wants all options to be open to the concretizer once you specify a package preference for a sticky variant.

Personally I think sticky variants always being sticky is fine, because if the user wants soft preference semantics for them in `packages.yaml`, they can put a list of values instead of a single value.  I wouldn't say that for regular variants, because I don't expect the user to list out all values of all variants.  But I *think* it's more intuitive to me if the stickiness is preserved.  You have to know the semantics of other variants to use them... why are these different?

But maybe a use case could convince me otherwise?